### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute '_run_module'

### ISSUE THREAD

**Issue by oapandit** [NVIDIA/NeMo-Guardrails#61]
Hi,

Thank you for creating this tool!

I am trying to run a simple rails which is already present in the code just to get started.
As I do not have OpenAI API key I was trying with huggingface model.

But while executing I am getting error from runhouse module. This error has originated at the time of 

**rails = LLMRails(config)**

### ERROR STACK

```typescript

File ~/softwares/nemo-guardrails/nemoguardrails/rails/llm/llmrails.py:79, in LLMRails.__init__(self, config, llm, verbose)
     76 self.runtime = Runtime(config=config, verbose=verbose)
     78 # Next, we initialize the LLM engine.
---> 79 self._init_llm()
     80 self.runtime.register_action_param("llm", self.llm)
     82 # Next, we initialize the LLM Generate actions and register them.

File ~/softwares/nemo-guardrails/nemoguardrails/rails/llm/llmrails.py:132, in LLMRails._init_llm(self)
    129         if hasattr(provider_cls, "model"):
    130             kwargs["model"] = main_llm_config.model
--> 132 self.llm = provider_cls(**kwargs)

File ~/miniconda3/envs/llm_eval/lib/python3.9/site-packages/langchain/llms/self_hosted_hugging_face.py:188, in SelfHostedHuggingFaceLLM.__init__(self, **kwargs)
    176 """Construct the pipeline remotely using an auxiliary function.
    177 
    178 The load function needs to be importable to be imported
    179 and run on the server, i.e. in a module and not a REPL or closure.
    180 Then, initialize the remote inference function.
    181 """
    182 load_fn_kwargs = {
    183     "model_id": kwargs.get("model_id", DEFAULT_MODEL_ID),
    184     "task": kwargs.get("task", DEFAULT_TASK),
    185     "device": kwargs.get("device", 0),
    186     "model_kwargs": kwargs.get("model_kwargs", None),
    187 }
--> 188 super().__init__(load_fn_kwargs=load_fn_kwargs, **kwargs)

File ~/miniconda3/envs/llm_eval/lib/python3.9/site-packages/langchain/llms/self_hosted.py:167, in SelfHostedPipeline.__init__(self, **kwargs)
    163 remote_load_fn = rh.function(fn=self.model_load_fn).to(
    164     self.hardware, reqs=self.model_reqs
    165 )
    166 _load_fn_kwargs = self.load_fn_kwargs or {}
--> 167 self.pipeline_ref = remote_load_fn.remote(**_load_fn_kwargs)
    169 self.client = rh.function(fn=self.inference_fn).to(
    170     self.hardware, reqs=self.model_reqs
    171 )

File ~/miniconda3/envs/llm_eval/lib/python3.9/site-packages/runhouse/rns/function.py:418, in Function.remote(self, *args, **kwargs)
    416 def remote(self, *args, **kwargs):
    417     warnings.warn("`remote()` is deprecated, use `run()` instead")
--> 418     run_obj = self.run(*args, **kwargs)
    419     return run_obj.name

File ~/miniconda3/envs/llm_eval/lib/python3.9/site-packages/runhouse/rns/function.py:441, in Function.run(self, run_name, *args, **kwargs)
    438 # Use the run_name if provided, otherwise create a new one using the Function's name
    439 run_name = run_name or Run._create_new_run_name(self.name)
--> 441 run_obj = self._call_fn_with_ssh_access(
    442     fn_type="remote", run_name=run_name, args=args, kwargs=kwargs
    443 )
    445 logger.info(f"Submitted remote call to cluster for {run_name}")
    446 return run_obj

File ~/miniconda3/envs/llm_eval/lib/python3.9/site-packages/runhouse/rns/function.py:492, in Function._call_fn_with_ssh_access(self, fn_type, resources, run_name, args, kwargs)
    489 if not isinstance(env_vars, dict):
    490     env_vars = _env_vars_from_file(env_vars)
--> 492 res = self.system._run_module(
    493     relative_path,
    494     module_name,
    495     fn_name,
    496     fn_type,
    497     resources,
    498     env_name,
    499     env_vars,
    500     run_name,
    501     args,
    502     kwargs,
    503 )
    504 return res

AttributeError: 'NoneType' object has no attribute '_run_module'
``` 

**Code snippet**

changed this in config.yml

```typescript

  - type: main
    engine: self_hosted_hugging_face
    model: decapoda-research/llama-7b-hf

``` 

models:

**Code to call LLM**

```python

from nemoguardrails import LLMRails, RailsConfig
import os

os.environ["HUGGINGFACEHUB_API_TOKEN"] = "HUGINGTOKEN"
def generate_message():
    # Give the path to the folder containing the rails
    config = RailsConfig.from_path("./examples/topical_rail")
    rails = LLMRails(config)

    # Define role and question to be asked
    new_message = rails.generate(messages=[{
        "role": "user",
        "content": "How can you help me?"
    }])
    print(new_message)


if __name__ == '__main__':
    generate_message()


``` 
I have tried to debug. It seems that in Runhouse library, the self.system variable is getting None value.
Can you please help in resolving this issue?

**Comment by drazvan** [comment]
Hi @oapandit!

It seems that initialising a `SelfHostedHuggingFaceLLM` requires more parameters (see https://github.com/hwchase17/langchain/blob/master/langchain/llms/self_hosted_hugging_face.py#L131). Specifically, the `hardware` parameter seems to be the one that ends up in the `system` local variable inside `runhouse`. 

In the YAML config from NeMo guardrails you can specify additional parameters using the `parameters` key (https://github.com/NVIDIA/NeMo-Guardrails/blob/main/docs/user_guide/configuration-guide.md#the-llm-model) but it works only with primitive types (string, int, etc.). Since the constructor for `SelfHostedHuggingFaceLLM` requires a `hardware` parameter, the only way to do it right now is to initialise the LLM manually. You can follow the example from here: https://github.com/NVIDIA/NeMo-Guardrails/blob/main/examples/llm/hf_pipeline_dolly/config.py and replace `get_dolly_v2_3b_llm()` with `get_self_hosted_hugging_face_llm()`. Inside the function you can initialise the LLM properly. 

Let me know if this works.

**Comment by oapandit** [comment]
Hello! @drazvan,

thanks for the answer.

I was unable to use selfhosted module so I have used huggingface pipeline way of loading my local models as those all are hf models. I followed similar setup as mentioned in https://github.com/NVIDIA/NeMo-Guardrails/blob/main/examples/llm/hf_pipeline_dolly/config.py and was able to make it work.

Maybe sometime I will try your suggestions to check if providing hardware variable value works.

thanks for your help!

best,
Onkar
